lib/portlayer/exec/container_cache.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"sync"
+)
+
+/*
+* ContainerCache will provide an in-memory cache of containerVMs.  It will
+* be refreshed on portlayer start and updated via container lifecycle
+* operations (start, stop, rm) and well as in response to infrastructure
+* events
+ */
+type containerCache struct {
+	m sync.RWMutex
+
+	// cache maps containerVM to container ID && infrastructure ID
+	cacheByID      map[string]*Container
+	cacheByInfraID map[string]*Container
+}
+
+var containers *containerCache
+
+func NewContainerCache() {
+	containers = &containerCache{
+		cacheByID:      make(map[string]*Container),
+		cacheByInfraID: make(map[string]*Container),
+	}
+}
+
+// returns a reference to a specific container
+//
+// idOrRef will either be a container ID or an infra
+// reference to the containerVM.  In the case of vSphere
+// the infra reference is a MoRef
+//
+func (cache *containerCache) Container(idOrRef string) *Container {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	// find by id
+	container := cache.cacheByID[idOrRef]
+	// if not found by id find by ref
+	if container == nil {
+		container = cache.cacheByInfraID[idOrRef]
+	}
+	return container
+}
+
+func (cache *containerCache) Add(container *Container) {
```

Comment:
This really should be called Put since we are overwriting what may be previously cached.